api/src/pcapi/celery_tasks/tasks.py

```diff
@@ -41,8 +87,25 @@ def task(payload: dict) -> None:
                     parsed_payload = model.parse_obj(parsed_payload)
                 except pydantic_v1.ValidationError as exp:
                     logger.error("could not deserialize object", extra={"exception": exp})
+            else:
+                parsed_payload = payload
 
-            f(parsed_payload)
+            if max_per_time_window:
+                try:
+                    with rate_limit(name, time_window_size, max_per_time_window):
+                        f(parsed_payload)
+                except RateLimitedError as exp:
+                    logger.error(
+                        "task rate-limited",
+                        extra={
+                            "time_window_size": exp.time_window_size,
+                            "max_per_time_window": exp.max_per_time_window,
+                            "current": exp.current,
+                        },
+                    )
+                    self.retry(exc=exp, countdown=time_window_size)
```

Comment:
Pour commencer, une bonne valeur de countdown pourrait tout simplement être le temps pour vider ceux de la file : 
```suggestion
                    self.retry(exc=exp, countdown=exp.time_window_size * (1 + (exp.current // exp.max_per_time_window))
```
Cela permettrait de ne pas tout réessayer dans la prochaine fenêtre si on sait déjà que ça ne passe pas.
Qu'en penses-tu ?

Reply:
Si je comprend bien tu propoes que si on a par exemple 10/s et qu'on a 30 tâches empilées, la 30e soit envoyé dans (3 * time_window_size) ?

Ca me paraît pas mal mais ça veut dire qu'il faut qu'on augmente le compteur de redis même quand on rate-limit(actuellement on le décompte pas).

Il faut peut-être aussi borner ça pour éviter de trop décaler dans le temps.

Reply:
J'ai implémenté ça du coup on garde le nombre de call au dessus de la limite et on s'en sert pour décider de dans combien de time_window on décalle. J'ai aussi mis une limite à 20 minutes et donc on prend le plus faible entre le résultat de ce calcul et la limite à 20mn